### ISSUE TITLE
Client not downloading .pkg files/It may have been corrupted or tampered with during downloading

### ISSUE THREAD

**Issue by tammala** [wdas/reposado#33]
Hello,

We are using reposado (ubuntu/nginx) to host updates to our OS X clients (10.10.3). I now have a problem when i try to install updates on a clients. 
When i try to install with managesofwateupdate, i get the following error message:

Installing iTunes...
	0    Error downloading iTunes: “iTunes” could not be verified. It may have been corrupted or tampered with during downloading. Click Update to download and check the update again.
    Done.
    Error downloading updates.
WARNING: Apple update iTunes, zzzz031-31871 failed to install. No record of success or failure.

I checked download location (Library/Managed Installs/swupd/content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut) on client and i noticed that all .pkg files are missing.

When i check my pilot branch catalog file from server, i can see that all files are listed. I can also download all missing pkg files with curl on client.

I checked nginx access log from server and noticed that all pkg files (+1 .dist file) are getting "Software%20Update (unknown version) CFNetwork/720.3.13 Darwin/14.3.0 (x86_64)".

10.0.72.77 - - [20/Aug/2015:09:34:29 +0300] "GET /index_pilot.sucatalog HTTP/1.1" 200 8877 "-" "managedsoftwareupdate/2.3.0.2519 Darwin/14.3.0 (x86_64) (iMac8,1)"
10.0.72.77 - - [20/Aug/2015:09:34:29 +0300] "GET /content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/031-31871.English.dist HTTP/1.1" 200 6644 "-" "Software%20Update (unknown version) CFNetwork/720.3.13 Darwin/14.3.0 (x86_64)"
10.0.72.77 - - [20/Aug/2015:09:34:30 +0300] "GET /content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/CoreADI.pkg HTTP/1.1" 200 1895738 "-" "Software%20Update (unknown version) CFNetwork/720.3.13 Darwin/14.3.0 (x86_64)"
10.0.72.77 - - [20/Aug/2015:09:34:30 +0300] "GET /content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/iTunesAccess.pkg HTTP/1.1" 200 687421 "-" "Software%20Update (unknown version) CFNetwork/720.3.13 Darwin/14.3.0 (x86_64)"
10.0.72.77 - - [20/Aug/2015:09:34:31 +0300] "GET /content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/MobileDevice.pkg HTTP/1.1" 200 32701834 "-" "Software%20Update (unknown version) CFNetwork/720.3.13 Darwin/14.3.0 (x86_64)"
10.0.72.77 - - [20/Aug/2015:09:34:31 +0300] "GET /content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/CoreFP.pkg HTTP/1.1" 200 26885432 "-" "Software%20Update (unknown version) CFNetwork/720.3.13 Darwin/14.3.0 (x86_64)"
10.0.72.77 - - [20/Aug/2015:09:34:34 +0300] "GET /content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/iTunesX.pkg HTTP/1.1" 200 164633993 "-" "Software%20Update (unknown version) CFNetwork/720.3.13 Darwin/14.3.0 (x86_64)"
1

These files were downloded ok.
10.0.72.77 - - [20/Aug/2015:09:51:53 +0300] "GET /content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/iTunesX.smd HTTP/1.1" 200 98472 "-" "managedsoftwareupdate/2.3.0.2519 Darwin/14.3.0 (x86_64) (iMac8,1)"
10.0.72.77 - - [20/Aug/2015:09:51:53 +0300] "GET /content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/CoreADI.pkm HTTP/1.1" 200 599 "-" "managedsoftwareupdate/2.3.0.2519 Darwin/14.3.0 (x86_64) (iMac8,1)"
10.0.72.77 - - [20/Aug/2015:09:51:53 +0300] "GET /content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/iTunesX.pkm HTTP/1.1" 200 2167 "-" "managedsoftwareupdate/2.3.0.2519 Darwin/14.3.0 (x86_64) (iMac8,1)"
1

Do you guys know what could cause this?

**Comment by arubdesu** [comment]
I'm certain I've heard this before, perhaps on the reposado or MacEnterprise google group mailing lists, but the 200's on the failed pkgs just means those are successful *delivered*. The local client is unhappy, which could mean something got messed in transit. You may want to confirm there are no proxies or other things that may be interfering with downloads either from or to your server. This is probably not a iTunes bug, since nobody else is reporting it, so I'd try a `repo_sync --recheck` before attempting things like purging the update and re-downloading.

**Comment by tammala** [comment]
This same problem happens with all my updates. We are using proxy server on clients but this should not be a problem, i tested manually and i can download pkg files from our server. I also tried to repo_sync recheck but not help. I also tested with apple default catalog->no help.

I was just wondering why computer type is not recognised on those .pkg files? pkm files are downloaded correctly.

**Comment by arubdesu** [comment]
Do the pkgs you're able to manually download via curl install properly? And the (unknown version) thing is a red herring, you'd see that in a caching server. Apple isn't sending identifiers for the softwareupdate app/binary/process. Good point, though, managedsoftwareupdate is munki, which uses the metadata and builds its own catalog, so I'd test with [`InstallAppleSoftwareUpdates` set to `False`](InstallAppleSoftwareUpdates) and see if it works with unmodified softwareupdate. What version of munki client are you using?

**Comment by tammala** [comment]
No, they don't install properly. There is same problem with regular softwareupdate. I'm using munki client 2.2.4.2431 but i also tested with 2.3.0.2519.
Do you know are there any other logs that i can check when download happens?

**Comment by gregneagle** [comment]
It sounds like you can remove Munki from the equation altogether: "There is same problem with regular softwareupdate"

" i tested manually and i can download pkg files from our server."
What does this mean, exactly? You have downloaded pkg files from your reposado server? If so, how have you then verified they are good?

It really sounds like you have corrupt pkg files in your reposado repo.

**Comment by tammala** [comment]
Something is really corrupting my reposado. I just re-synced everything and still same problem.
I just tried to download same file from my reposado server and apple. I couldn't install iTunesX.pkg that was downloaded from my server. Same package, same size and etc. I have to talk with network guys if they something from firewall logs.

curl -O http://servername.edu.xxx.fi:8088/content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/iTunesX.pkg

curl -O http://swcdn.apple.com/content/downloads/34/21/zzzz031-31871/oiafge5g10rnspc2kdm3qusk1dky3jbhut/iTunesX.pkg

**Comment by jessepeterson** [comment]
@tammala: What's a hash of those files you've downloaded? I get an MD5 of `eb4db7f28f5dfdf905c83923a949a14f` on iTunesX.pkg from swcdn.apple.com.

**Comment by tammala** [comment]
@jessepeterson  I checked hashes and it was different (dd754dea16e2be3536000f96dcaebafa). I then downloaded same iTunesX.pkg (from apple) with my admin computer, checked hashes and it was same as yours.

I then added our proxy server to reposado preferences file and re-downloaded iTunesX.pkg. Now i get same hash than you also on my reposado server.
So something on my network is really corrupting pkgs. Proxy server is not required on our server subnet, it's only required on our client subnets.

I then deleted all catalogs/downloads on reposado server and did reposync. Now it seems that everything is working as expected. I have to talk with network guys and check what could cause this.

Not working:
MD5 (iTunesX.pkg) = dd754dea16e2be3536000f96dcaebafa
Working:
MD5 (iTunesX.pkg) = eb4db7f28f5dfdf905c83923a949a14f
